src/gar.lua

```diff
@@ -1577,4 +1577,271 @@ function gar.getPaginatedDelegations(address, cursor, limit, sortBy, sortOrder)
+			)
+
+			-- Check if the delegate will have enough stake left after re-delegating
+			local existingStake = existingDelegate.delegatedStake
+			local requiredMinimumStake = sourceGateway.settings.minDelegatedStake
+			local maxAllowedToWithdraw = existingStake - requiredMinimumStake
+			assert(
+				stakeToTakeFromSource <= maxAllowedToWithdraw or stakeToTakeFromSource == existingStake,
+				"Remaining delegated stake must be greater than the minimum delegated stake. Adjust the amount or re-delegate all stake."
+			)
+
+			-- If the delegate has enough stake to redelegate, move the stake. If its all the stake, remove the delegate
+			if existingDelegate.delegatedStake == stakeToTakeFromSource then
+				gar.pruneDelegateFromGateway(delegateAddress, sourceGateway)
+			else
+				sourceGateway.delegates[delegateAddress].delegatedStake = sourceGateway.delegates[delegateAddress].delegatedStake
+					- stakeToTakeFromSource
+			end
+			sourceGateway.totalDelegatedStake = sourceGateway.totalDelegatedStake - stakeToTakeFromSource
+		end
+	end
+
+	local existingTargetDelegate = targetGateway.delegates[delegateAddress]
+	local minimumStakeForGatewayAndDelegate
+	if existingTargetDelegate and existingTargetDelegate.delegatedStake ~= 0 then
+		-- It already has a stake that is not zero
+		minimumStakeForGatewayAndDelegate = 1 -- Delegate must provide at least one additional IO
+	else
+		-- Consider if the operator increases the minimum amount after you've already staked
+		minimumStakeForGatewayAndDelegate = targetGateway.settings.minDelegatedStake
+	end
+
+	-- Check if the delegate has enough stake to redelegate
+	assert(
+		stakeToDelegate >= minimumStakeForGatewayAndDelegate,
+		"Quantity must be greater than the minimum delegated stake amount."
+	)
+
+	-- The stake can now be applied to the targetGateway
+	if targetAddress == delegateAddress then
+		-- move the stake to the operator's stake
+		targetGateway.operatorStake = targetGateway.operatorStake + stakeToDelegate
+	else
+		local previousDelegate = targetGateway.delegates[delegateAddress]
+		local previousDelegateStake = previousDelegate and previousDelegate.delegatedStake or 0
+		local previousDelegateVaults = previousDelegate and previousDelegate.vaults or {}
+		local previousDelegateStartTimestamp = previousDelegate and previousDelegate.startTimestamp or currentTimestamp
+		targetGateway.delegates[delegateAddress] = {
+			delegatedStake = previousDelegateStake + stakeToDelegate,
+			startTimestamp = previousDelegateStartTimestamp,
+			vaults = previousDelegateVaults,
+		}
+		targetGateway.totalDelegatedStake = targetGateway.totalDelegatedStake + stakeToDelegate
+	end
+
+	-- Move redelegation fee to protocol balance
+	balances.increaseBalance(ao.id, redelegationFee)
+
+	Redelegations[delegateAddress] = {
+		timestamp = currentTimestamp,
+		redelegations = (previousRedelegations and previousRedelegations.redelegations or 0) + 1,
+	}
+
+	-- prune user from allow list, if necessary, to save memory
```

Comment:
Not quite how this works. User goes back on the allowlist on the source gateway IF they've been pruned, otherwise their active balance or vaults "holds their spot" on the gateway. On the destination gateway, we take them out of the allowlist lookup if we create a new delegation for them.

Reply:
Standardized this with a few utility functions in the gar. Should be all good now.